design/authorization.md

```diff
@@ -0,0 +1,85 @@
+
+
+## Authorization scope
+It is important to remember that Nessie does not store data directly but only data location and other metadata. As a consequence Nessie authorization layer can only really control access to metadata, but might not prevent data itself to be accessed directly without interacting with Nessie. It is then expected that another system can control access to data itself to make sure no authorized access is made.
+The same is true for access to historical data which is one of Nessie main features. For example while it might seem safe when detecting some undesired sensitive data to commit a change to remove it and to restrict access to only the latest version of the dataset, the truth is that the sensitive data may still exist on the datalake and be accessed by other means (similar to how redacting a PDF by adding black boxes on top of sensitive information does not prevent people to read what is written beneath in most cases). The only safe method to remove this data is to remove it from the datalake, and to add commits to Nessie to point to the new data location.
+
+## Stories
+Here's a list of common authorization scenarios:
+
+* Alice attempts to execute a query against the table `Foo` on branch `prod`. As she has read access to the table on this branch, Nessie allows the execution engine to get the table details so it can locate the data and try to access it.
+* Bob attempts to execute a query against the table `Foo` on branch `prod`. However Bob does not have read access to the table. Nessie returns an authorization error, and the execution engine refuses to execute the query.
+* Carol has access to the branch `prod` content, but not to the table `Foo` on this branch. Carol creates a new reference named `carol-branch`with the same hash as `prod`, and attempts to change permission on table `Foo`. However request is denied, and Carol cannot access the content of `Foo`.
+* Dave has access to the branch `prod` content, and wants to update the content of the table `Foo`. He creates a new reference named `dave-experiment`, and executes several queries against this branch to modify table `Foo`. Each modification is a commit done against `dave-experiment` branch which is approved by Nessie server. When all the desired modifications are done, Dave attempts to merge the changes back to the `prod` branch. However Dave doesn't have the rights to modify the `prod` branch, causing Nessie to deny the request.
+
+## Access control model
+Any object in Nessie can be designated by a pair of coordinates (reference, path). It seems logical that access control is also designed around those two concepts
+
+### Reference access control
+References can be designated by name or by hash. However the practicality of using hashes for access control seems very low as those hashes are supposedly random and not known in advance. So focus will be on named references (branches and tags).
```

Comment:
what happens if we ask for (key, hash) would it always accept/always deny? Eg the Iceberg client does a minimum of isolation control by only asking for hashes (as opposed to branch/tag). Is it possible to have a hash which is a parent of two branches where the branches have different permissions for a key?